File: src/javascript/algorithms/graph/networkDelayTime/networkDelayTime.js
// Network Delay Time - Dijkstra's Algorithm
function networkDelayTime(times, n, k) {
  const graph = {};
  for (const [u, v, w] of times) {
    if (!graph[u]) graph[u] = [];
    graph[u].push([v, w]);
  }
  const minHeap = [[0, k]];
  const dist = {};
  while (minHeap.length) {
    minHeap.sort((a, b) => a[0] - b[0]);
    const [time, node] = minHeap.shift();
    if (dist[node] !== undefined) continue;
    dist[node] = time;
    if (graph[node]) {
      for (const [nei, wt] of graph[node]) {
        if (dist[nei] === undefined) {
          minHeap.push([time + wt, nei]);
        }
      }
    }
  }
  return Object.keys(dist).length === n ? Math.max(...Object.values(dist)) : -1;
}

// Test Cases
console.log(networkDelayTime([[2,1,1],[2,3,1],[3,4,1]], 4, 2)); // 2
console.log(networkDelayTime([[1,2,1]], 2, 1)); // 1
console.log(networkDelayTime([[1,2,1]], 2, 2)); // -1
console.log(networkDelayTime([[1,2,1],[2,3,2]], 4, 1)); // -1

File: src/python/algorithms/graph/networkDelayTime/network_delay_time.py
# Network Delay Time - Dijkstra's Algorithm

def network_delay_time(times, n, k):
    import heapq
    from collections import defaultdict
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))
    min_heap = [(0, k)]
    dist = {}
    while min_heap:
        time, node = heapq.heappop(min_heap)
        if node in dist:
            continue
        dist[node] = time
        for nei, wt in graph[node]:
            if nei not in dist:
                heapq.heappush(min_heap, (time + wt, nei))
    return max(dist.values()) if len(dist) == n else -1

if __name__ == "__main__":
    # Test Case 1
    print(network_delay_time([[2,1,1],[2,3,1],[3,4,1]], 4, 2)) # 2
    # Test Case 2
    print(network_delay_time([[1,2,1]], 2, 1)) # 1
    # Test Case 3
    print(network_delay_time([[1,2,1]], 2, 2)) # -1
    # Test Case 4
    print(network_delay_time([[1,2,1],[2,3,2]], 4, 1)) # -1
